### PR TITLE
Problem: Unable to modify sb_threshold at runtime

### LINKED ISSUE #368: Problem: Cannot update sb_threshold value
Currently the sb_threshold can only be set initially when a socket is created via `usrsctp_socket`. We should add a function to be able to update this value.

A possible use case would be to update the `sb_threshold` size from within the `send_callback` to avoid spurious wakeups.

I'll write a PR.

### PR DESCRIPTION
**Pull request by jean-airoldie**
Solution: Add usrsctp_set_sb_threshold function.

Closes #368.

### CODE REVIEW

File: usrsctplib/netinet/sctp_usrreq.c [hunk 1]
```diff
@@ -9073,4 +9073,19 @@ register_ulp_info (struct socket *so, void *ulp_info)
 	SCTP_INP_WUNLOCK(inp);
 	return (1);
 }
+int
+
+register_sb_threshold(struct socket *so, uint32_t sb_threshold)
+{
+	struct sctp_inpcb *inp;
+
```
Comment (tuexen) [inline usrsctplib/netinet/sctp_usrreq.c]: Can you add:
```
	if (so == NULL) {
		errno = EBADF;
		return (-1);
	}
```
as it is checked in most other functions?

File: usrsctplib/user_socket.c [hunk 2]
```diff
@@ -2584,6 +2584,12 @@ usrsctp_set_ulpinfo(struct socket *so, void *ulp_info)
 	return (register_ulp_info(so, ulp_info));
 }
 
+int
+usrsctp_set_sb_threshold(struct socket *so, uint32_t sb_threshold)
+{
```
Comment (tuexen) [inline usrsctplib/user_socket.c]: Can you add
```
	if (so == NULL) {
		errno = EBADF;
		return (-1);
	}
```
as it is done for most other functions (and missing from `usrsctp_set_ulpinfo()`?

File: usrsctplib/usrsctp.h [hunk 3]
```diff
@@ -1030,6 +1030,9 @@ usrsctp_deregister_address(void *);
 int
 usrsctp_set_ulpinfo(struct socket *, void *);
 
+int
+usrsctp_set_sb_threshold(struct socket *, uint32_t sb_threshold);
```
Comment (tuexen) [inline usrsctplib/usrsctp.h]: Please remove `sb_threshold` for consistency.

**Comment by jean-airoldie** [comment]
Now that I think about it, shouldn't `sb_threshold` be set on a association basis as there is a send_buffer for each association (from what i understand)? We would also need to add another `assoc_id` field to the `send_callback`.

**Comment by jean-airoldie** [comment]
Otherwise a stalling association could potentially block all other associations.

**Comment by jean-airoldie** [comment]
We would need to add the threshold field directly to the https://github.com/sctplab/usrsctp/blob/8ca132191668c1dff35872315cf399cd7c3dba9b/usrsctplib/netinet/sctp_pcb.h#L614
since it would be on an association basis.

**Comment by jean-airoldie** [comment]
I'll close this issue since its too tightly linked with #367.

**Comment by jean-airoldie** [comment]
Alright so I think the `one-to-many` API wouldn't work for my needs because I cannot chose from which association I recv a message using `recvv`. However, the `one-to-one` API could work.

Moreover, in the `one-to-one` API the `sb_threshold` acts as the threshold for that association, so we don't run into the problem I previously raised.

File: usrsctplib/usrsctp.h (hunk 3 above)
Comment (jean-airoldie) [reply usrsctplib/usrsctp.h]: Done.

Comment (tuexen) [reply usrsctplib/usrsctp.h]: Really?

Comment (jean-airoldie) [reply usrsctplib/usrsctp.h]: Right sorry, merged with the upstream.

**Comment by jean-airoldie** [comment]
I'm no lounger using the callback API, in favor of the upcall API, so there is no point to this PR.

**Pull request closed without merging**